builder/cse.cc

```diff
@@ -0,0 +1,547 @@
+      include(l.first);
+      include(l.second);
+    }
+    include(op->body);
+  }
+  void visit(const block* op) override {
+    for (const auto& s : op->stmts) {
+      include(s);
+    }
+  }
+  void visit(const loop* op) override {
+    include(op->sym);
+    include(op->bounds);
+    include(op->step);
+    include(op->body);
+  }
+  void visit(const call_stmt* op) override {}
+  void visit(const copy_stmt* op) override {}
+  void visit(const allocate* op) override {
+    include(op->sym);
+    include(op->elem_size);
+    for (const auto& d : op->dims) {
+      include(d);
+    }
+    include(op->body);
+  }
+  void visit(const make_buffer* op) override {
+    include(op->base);
+    include(op->elem_size);
+    for (const auto& d : op->dims) {
+      include(d);
+    }
+    include(op->body);
+  }
+  void visit(const clone_buffer* op) override {
+    include(op->sym);
+    include(op->src);
+    include(op->body);
+  }
+  void visit(const crop_buffer* op) override {
+    include(op->sym);
+    include(op->src);
+    for (const auto& b : op->bounds) {
+      include(b);
+    }
+    include(op->body);
+  }
+  void visit(const crop_dim* op) override {
+    include(op->sym);
+    include(op->src);
+    include(op->bounds);
+    include(op->body);
+  }
+  void visit(const slice_buffer* op) override {
+    include(op->sym);
+    include(op->src);
+    for (const auto& a : op->at) {
+      include(a);
+    }
+    include(op->body);
+  }
+  void visit(const slice_dim* op) override {
+    include(op->sym);
+    include(op->src);
+    include(op->at);
+    include(op->body);
+  }
+  void visit(const transpose* op) override {
+    include(op->sym);
+    include(op->src);
+    for (const auto& d : op->dims) {
+      include(d);
+    }
+    include(op->body);
+  }
+  void visit(const check* op) override { include(op->condition); }
+
+private:
+  // The nodes visited so far
+  std::set<const void*> visited;
+};
+
+// A mutator that caches and reapplies previously-done mutations, so
+// that it can handle graphs of IR that have not had CSE done to
+// them.
+//
+// TODO: should this be moved into node_mutator.h?
+class graph_node_mutator : public node_mutator {
+protected:
+  std::map<expr, expr, expr_compare_shallow> expr_replacements;
+  std::map<stmt, stmt, stmt_compare_shallow> stmt_replacements;
+
+public:
+  stmt mutate(const stmt& s) override {
+    auto p = stmt_replacements.emplace(s, stmt());
+    if (p.second) {
+      // N.B: Inserting into a map (as the recursive mutate call
+      // does), does not invalidate existing iterators.
+      p.first->second = node_mutator::mutate(s);
+    }
+    return p.first->second;
+  }
+
+  expr mutate(const expr& e) override {
+    auto p = expr_replacements.emplace(e, expr());
+    if (p.second) {
+      p.first->second = node_mutator::mutate(e);
+    }
+    return p.first->second;
+  }
+};
+
+// ----------------------
+
+bool is_const(const expr& e) { return e.as<constant>() != nullptr; }
```

Comment:
We already have `as_constant` and `as_variable` which can serve as these functions.